File: TPoint/TPoint/Presentation/ViewModels/GameViewModel.swift
//
//  GameViewModel.swift
//  TPoint
//

import Foundation

protocol GameViewModelDelegate: AnyObject {
    func gameDidUpdate()
    func gameDidEnd(winner: TeamType)
}

enum ThemeMode: Int {
    case system = 0
    case light = 1
    case dark = 2
}

final class GameViewModel {

    // MARK: - UserDefaults Keys

    private enum Keys {
        static let targetScore = "targetScore"
        static let themeMode = "themeMode"
    }

    // MARK: - Properties

    weak var delegate: GameViewModelDelegate?

    private(set) var game: Game
    private let calculateScoreUseCase: CalculateScoreUseCaseProtocol
    private let storage = GameStorage()
    private let historyStorage = HistoryStorage()

    // 현재 라운드 입력 상태
    var currentTeamACardScore: Int = 50
    var currentOneTwoFinish: Bool = false
    var currentOneTwoFinishTeam: TeamType?
    var currentTichuCalls: [Player: TichuCallInput] = [:]

    // 설정 - 목표 점수
    var targetScore: Int {
        get {
            let saved = UserDefaults.standard.integer(forKey: Keys.targetScore)
            return saved > 0 ? saved : 1000
        }
        set {
            UserDefaults.standard.set(newValue, forKey: Keys.targetScore)
            game.targetScore = newValue
            persist()
            delegate?.gameDidUpdate()
        }
    }

    // 설정 - 테마 모드
    var themeMode: ThemeMode {
        get {
            ThemeMode(rawValue: UserDefaults.standard.integer(forKey: Keys.themeMode)) ?? .system
        }
        set {
            UserDefaults.standard.set(newValue.rawValue, forKey: Keys.themeMode)
        }
    }

    // MARK: - Computed Properties

    var teamAScore: Int { game.teamA.totalScore }
    var teamBScore: Int { game.teamB.totalScore }
    var rounds: [Round] { game.rounds }
    var roundCount: Int { game.rounds.count }
    var isGameOver: Bool { game.isGameOver }
    var winner: TeamType? { game.winner }

    var currentRound: Int { game.rounds.count + 1 }

    var currentTeamBCardScore: Int {
        100 - currentTeamACardScore
    }

    // MARK: - Initialization

    init(calculateScoreUseCase: CalculateScoreUseCaseProtocol = CalculateScoreUseCase()) {
        self.game = Game()
        self.calculateScoreUseCase = calculateScoreUseCase
        // Load saved target score
        let savedTargetScore = UserDefaults.standard.integer(forKey: Keys.targetScore)
        if savedTargetScore > 0 {
            self.game.targetScore = savedTargetScore
        }
        resetCurrentRoundInput()
        // 저장된 진행 중 게임이 있으면 복원
        restoreIfAvailable()
    }

    // MARK: - Persistence

    private func restoreIfAvailable() {
        guard let snapshot = storage.load() else { return }
        game = snapshot.game
        currentTeamACardScore = snapshot.currentTeamACardScore
        currentOneTwoFinish = snapshot.currentOneTwoFinish
        currentOneTwoFinishTeam = snapshot.currentOneTwoFinishTeam
        currentTichuCalls = snapshot.currentTichuCalls
    }

    private func persist() {
        storage.save(GameSnapshot(
            game: game,
            currentTeamACardScore: currentTeamACardScore,
            currentOneTwoFinish: currentOneTwoFinish,
            currentOneTwoFinishTeam: currentOneTwoFinishTeam,
            currentTichuCalls: currentTichuCalls
        ))
    }

    // MARK: - Public Methods

    func setTeamACardScore(_ score: Int) {
        // 카드 점수는 -25~125 범위 (슬라이더 및 Android와 동일). 상대팀은 100 - 값.
        currentTeamACardScore = max(-25, min(125, score))
        persist()
        delegate?.gameDidUpdate()
    }

    func setOneTwoFinish(enabled: Bool, team: TeamType?) {
        currentOneTwoFinish = enabled
        currentOneTwoFinishTeam = team
        persist()
        delegate?.gameDidUpdate()
    }

    func setTichuCall(for player: Player, type: TichuType?, isSuccess: Bool) {
        if let type = type {
            currentTichuCalls[player] = TichuCallInput(type: type, isSuccess: isSuccess)
        } else {
            currentTichuCalls.removeValue(forKey: player)
        }
        persist()
        delegate?.gameDidUpdate()
    }

    func getTichuCall(for player: Player) -> TichuCallInput? {
        currentTichuCalls[player]
    }

    func addRound() {
        // 원투피니시 시 상대팀은 1등이 불가능하므로 티추 성공을 강제로 실패 처리
        let oneTwoOpponent = currentOneTwoFinish ? currentOneTwoFinishTeam?.opponent : nil
        let tichuCalls = currentTichuCalls.compactMap { (player, input) -> TichuCall? in
            let isSuccess = (player.team == oneTwoOpponent) ? false : input.isSuccess
            return TichuCall(
                player: player,
                isLarge: input.type == .large,
                isSuccess: isSuccess
            )
        }

        let round = Round(
            roundNumber: game.rounds.count + 1,
            teamACardScore: currentTeamACardScore,
            isOneTwoFinish: currentOneTwoFinish,
            oneTwoFinishTeam: currentOneTwoFinishTeam,
            tichuCalls: tichuCalls
        )

        let score = calculateScoreUseCase.calculate(round: round)

        game.rounds.append(round)
        game.teamA.totalScore += score.teamAScore
        game.teamB.totalScore += score.teamBScore

        resetCurrentRoundInput()
        persist()

        if let winner = game.winner {
            delegate?.gameDidEnd(winner: winner)
        } else {
            delegate?.gameDidUpdate()
        }
    }

    func getRoundScore(at index: Int) -> RoundScore? {
        guard index < game.rounds.count else { return nil }
        return calculateScoreUseCase.calculate(round: game.rounds[index])
    }

    func newGame() {
        // 라운드가 있는 현재 게임은 새 게임 시작 시 기록에 보관
        if !game.rounds.isEmpty {
            historyStorage.append(GameHistoryEntry(completedAt: Date(), game: game))
        }
        game.reset()
        resetCurrentRoundInput()
        persist()
        delegate?.gameDidUpdate()
    }

    func loadHistory() -> [GameHistoryEntry] {
        historyStorage.load()
    }

    func clearHistory() {
        historyStorage.clear()
    }

    func deleteRound(at index: Int) {
        guard index >= 0 && index < game.rounds.count else { return }

        let round = game.rounds[index]
        let score = calculateScoreUseCase.calculate(round: round)

        // 점수 차감
        game.teamA.totalScore -= score.teamAScore
        game.teamB.totalScore -= score.teamBScore

        // 라운드 삭제
        game.rounds.remove(at: index)

        // 라운드 번호 재정렬
        for i in index..<game.rounds.count {
            game.rounds[i] = Round(
                roundNumber: i + 1,
                teamACardScore: game.rounds[i].teamACardScore,
                isOneTwoFinish: game.rounds[i].isOneTwoFinish,
                oneTwoFinishTeam: game.rounds[i].oneTwoFinishTeam,
                tichuCalls: game.rounds[i].tichuCalls
            )
        }

        persist()
        delegate?.gameDidUpdate()
    }

    // MARK: - Private Methods

    private func resetCurrentRoundInput() {
        currentTeamACardScore = 50
        currentOneTwoFinish = false
        currentOneTwoFinishTeam = nil
        currentTichuCalls = [:]
    }
}

// MARK: - Supporting Types

enum TichuType: String, Codable {
    case small  // 100점
    case large  // 200점
}

struct TichuCallInput: Codable {
    var type: TichuType
    var isSuccess: Bool
}

File: TichuTallyAndroid/app/src/main/java/com/tichutally/app/presentation/viewmodel/GameViewModel.kt
package com.tichutally.app.presentation.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.tichutally.app.data.GameHistoryEntry
import com.tichutally.app.data.GameStorage
import com.tichutally.app.data.HistoryStorage
import com.tichutally.app.domain.model.*
import com.tichutally.app.domain.usecase.CalculateScoreUseCase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class GameViewModel(application: Application) : AndroidViewModel(application) {

    private val calculateScoreUseCase = CalculateScoreUseCase()
    private val storage = GameStorage(application)
    private val historyStorage = HistoryStorage(application)

    // 저장된 진행 중 게임이 있으면 복원, 없으면 새 게임으로 시작
    private val _state = MutableStateFlow(storage.load() ?: GameState())
    val state: StateFlow<GameState> = _state.asStateFlow()

    @OptIn(FlowPreview::class)
    private val autoSave = viewModelScope.launch(Dispatchers.IO) {
        // 상태 변경 시 자동 저장 (연속 슬라이더 입력은 디바운스로 합침)
        state.drop(1).debounce(250).collect { storage.save(it) }
    }

    fun setTeamACardScore(score: Int) {
        _state.update { it.copy(currentTeamACardScore = score.coerceIn(-25, 125)) }
    }

    fun setOneTwoFinish(enabled: Boolean, team: TeamType?) {
        _state.update {
            it.copy(
                currentOneTwoFinish = enabled,
                currentOneTwoFinishTeam = team
            )
        }
    }

    fun setTichuCall(player: Player, type: TichuType?, isSuccess: Boolean) {
        _state.update { currentState ->
            val newCalls = currentState.currentTichuCalls.toMutableMap()

            if (type != null) {
                newCalls[player] = TichuCallInput(type, isSuccess)

                // Auto-fail other tichu calls when one succeeds
                if (isSuccess) {
                    Player.allPlayers.filter { it != player }.forEach { otherPlayer ->
                        newCalls[otherPlayer]?.let { call ->
                            if (call.type != null) {
                                newCalls[otherPlayer] = call.copy(isSuccess = false)
                            }
                        }
                    }
                }
            } else {
                newCalls.remove(player)
            }

            currentState.copy(currentTichuCalls = newCalls)
        }
    }

    fun addRound() {
        _state.update { currentState ->
            // 원투피니시 시 상대팀은 1등이 불가능하므로 티추 성공을 강제로 실패 처리
            val oneTwoOpponent = if (currentState.currentOneTwoFinish)
                currentState.currentOneTwoFinishTeam?.opponent else null
            val tichuCalls = currentState.currentTichuCalls.mapNotNull { (player, input) ->
                input.type?.let { type ->
                    val success = if (player.team == oneTwoOpponent) false else input.isSuccess
                    TichuCall(player, type, success)
                }
            }

            val round = Round(
                roundNumber = currentState.game.rounds.size + 1,
                teamACardScore = currentState.currentTeamACardScore,
                isOneTwoFinish = currentState.currentOneTwoFinish,
                oneTwoFinishTeam = currentState.currentOneTwoFinishTeam,
                tichuCalls = tichuCalls
            )

            val score = calculateScoreUseCase.calculate(round)

            val newGame = currentState.game.copy(
                teamA = currentState.game.teamA.copy(
                    totalScore = currentState.game.teamA.totalScore + score.teamAScore
                ),
                teamB = currentState.game.teamB.copy(
                    totalScore = currentState.game.teamB.totalScore + score.teamBScore
                ),
                rounds = currentState.game.rounds + round
            )

            val newRoundScores = currentState.roundScores + score

            currentState.copy(
                game = newGame,
                roundScores = newRoundScores,
                currentTeamACardScore = 50,
                currentOneTwoFinish = false,
                currentOneTwoFinishTeam = null,
                currentTichuCalls = emptyMap(),
                showWinnerDialog = newGame.isGameOver
            )
        }
    }

    fun dismissWinnerDialog() {
        _state.update { it.copy(showWinnerDialog = false) }
    }

    fun setTargetScore(score: Int) {
        _state.update { currentState ->
            val wasOver = currentState.game.isGameOver
            val newGame = currentState.game.copy(targetScore = score)
            // 게임이 새로 종료된 경우에만 승자 다이얼로그 노출.
            // 이미 종료돼 있었다면 기존 노출 상태 유지(닫아둔 걸 다시 띄우지 않음), 해제되면 숨김.
            val showWinner = newGame.isGameOver && (currentState.showWinnerDialog || !wasOver)
            currentState.copy(
                targetScore = score,
                game = newGame,
                showWinnerDialog = showWinner
            )
        }
    }

    fun setThemeMode(mode: ThemeMode) {
        _state.update { it.copy(themeMode = mode) }
    }

    fun setTeamNames(teamAName: String, teamBName: String) {
        _state.update { it.copy(teamAName = teamAName.trim(), teamBName = teamBName.trim()) }
    }

    fun newGame() {
        // 라운드가 있는 현재 게임은 새 게임 시작 시 기록에 보관
        val current = _state.value
        if (current.game.rounds.isNotEmpty()) {
            val entry = GameHistoryEntry(System.currentTimeMillis(), current.game)
            viewModelScope.launch(Dispatchers.IO) { historyStorage.append(entry) }
        }
        _state.update { currentState ->
            GameState(
                targetScore = currentState.targetScore,
                themeMode = currentState.themeMode,
                teamAName = currentState.teamAName,
                teamBName = currentState.teamBName,
                game = Game(targetScore = currentState.targetScore)
            )
        }
    }

    fun loadHistory(): List<GameHistoryEntry> = historyStorage.load()

    fun clearHistory() {
        viewModelScope.launch(Dispatchers.IO) { historyStorage.clear() }
    }

    fun updateRound(
        index: Int,
        teamACardScore: Int,
        isOneTwoFinish: Boolean,
        oneTwoFinishTeam: TeamType?,
        tichuCalls: Map<Player, TichuCallInput>
    ) {
        _state.update { currentState ->
            val rounds = currentState.game.rounds
            if (index < 0 || index >= rounds.size) return@update currentState
            val oldScore = currentState.roundScores.getOrNull(index) ?: return@update currentState

            // 원투피니시 시 상대팀 티추 성공 강제 실패
            val oneTwoOpponent = if (isOneTwoFinish) oneTwoFinishTeam?.opponent else null
            val calls = tichuCalls.mapNotNull { (player, input) ->
                input.type?.let { type ->
                    val success = if (player.team == oneTwoOpponent) false else input.isSuccess
                    TichuCall(player, type, success)
                }
            }
            val newRound = rounds[index].copy(
                teamACardScore = teamACardScore.coerceIn(-25, 125),
                isOneTwoFinish = isOneTwoFinish,
                oneTwoFinishTeam = oneTwoFinishTeam,
                tichuCalls = calls
            )
            val newScore = calculateScoreUseCase.calculate(newRound)

            val newRounds = rounds.toMutableList().apply { set(index, newRound) }
            val newRoundScores = currentState.roundScores.toMutableList().apply { set(index, newScore) }
            val newTeamA = currentState.game.teamA.copy(
                totalScore = currentState.game.teamA.totalScore - oldScore.teamAScore + newScore.teamAScore
            )
            val newTeamB = currentState.game.teamB.copy(
                totalScore = currentState.game.teamB.totalScore - oldScore.teamBScore + newScore.teamBScore
            )
            val wasOver = currentState.game.isGameOver
            val newGame = currentState.game.copy(teamA = newTeamA, teamB = newTeamB, rounds = newRounds)
            currentState.copy(
                game = newGame,
                roundScores = newRoundScores,
                showWinnerDialog = newGame.isGameOver && (currentState.showWinnerDialog || !wasOver)
            )
        }
    }

    // 삭제 직전 상태 (실행취소용)
    private var undoSnapshot: GameState? = null

    fun deleteRound(index: Int) {
        val pre = _state.value
        val rounds = pre.game.rounds
        if (index < 0 || index >= rounds.size) return
        val deletedScore = pre.roundScores.getOrNull(index) ?: return

        // 실행취소를 위해 삭제 직전 상태 보관 (플래그는 제외)
        undoSnapshot = pre.copy(showUndoDelete = false)

        val newTeamA = pre.game.teamA.copy(totalScore = pre.game.teamA.totalScore - deletedScore.teamAScore)
        val newTeamB = pre.game.teamB.copy(totalScore = pre.game.teamB.totalScore - deletedScore.teamBScore)

        // 라운드 삭제 및 번호 재정렬
        val newRounds = rounds.toMutableList().apply { removeAt(index) }
            .mapIndexed { i, round -> round.copy(roundNumber = i + 1) }
        val newRoundScores = pre.roundScores.toMutableList().apply { removeAt(index) }

        _state.value = pre.copy(
            game = pre.game.copy(teamA = newTeamA, teamB = newTeamB, rounds = newRounds),
            roundScores = newRoundScores,
            showWinnerDialog = false,
            showUndoDelete = true
        )
    }

    fun undoDelete() {
        undoSnapshot?.let { _state.value = it }
        undoSnapshot = null
    }

    fun dismissUndo() {
        undoSnapshot = null
        _state.update { it.copy(showUndoDelete = false) }
    }
}
